doc/Running and Writing Tests for Sia.md

```diff
@@ -0,0 +1,278 @@
+// database:
+// - a database that has not yet been created
+// - an existing empty database
+// - an existing nonempty database
+// Along the way, it also tests calling Close on:
+// - a newly-created database
+// - a newly-filled database
+// - a newly-emptied database
+func TestOpenDatabase(t *testing.T) {
+    // Since the test involves writing a bunch of files to disk, skip it when
+    // `go test` is called with the `-short` flag
+	if testing.Short() {
+		t.SkipNow()
+	}
+
+    // Declare testInputs locally since TestOpenDatabase is the only function
+    // using them.
+	testInputs := []struct {
+		dbMetadata Metadata
+		dbFilename string
+	}{
+		{Metadata{"", ""}, " "},
+		{Metadata{"", ""}, "_"},
+		{Metadata{"_", "_"}, "_"},
+		{Metadata{"asdf", "asdf"}, "asdf"},
+		{Metadata{"1sadf23", "12253"}, "123kjhgfd"},
+		{Metadata{"$@#$%^&", "$@#$%^&"}, "$@#$%^&"},
+		{Metadata{"//", "//"}, "_"},
+		{Metadata{"testHeader" + RandomSuffix(), "0.0.0"}, "testFilename" + RandomSuffix()},
+		{Metadata{"testHeader" + RandomSuffix(), "7.0.4"}, "testFilename" + RandomSuffix()},
+		{Metadata{"testHeader" + RandomSuffix(), "asdf"}, "testFilename" + RandomSuffix()},
+		{Metadata{"testHeader" + RandomSuffix(), ""}, "testFilename" + RandomSuffix()},
+		{Metadata{"testHeader" + RandomSuffix(), "_"}, "testFilename" + RandomSuffix()},
+		{Metadata{"I am he as you are he as you are me And we are all together See how they run like pigs from a gun see how they fly I'm crying", "Sitting on a cornflake waiting for the van to come Corporation tee shirt, stupid bloody Tuesday Man you been a naughty boy. You let your face grow long I am the eggman, they are the eggmen I am the walrus, goo goo g' joob"}, "Mister City Policeman sitting, pretty little policemen in a row See how they fly like Lucy in the sky, see how they run I'm crying, I'm crying I'm crying, I'm crying"},
+		{Metadata{":]", ":)"}, ":|"},
+		{Metadata{"¯|_(ツ)_|¯", "can't do the usual arms because the computer gets confused"}, "¯|_(ツ)_|¯"},
+		{Metadata{"世界", "怎么办呢"}, "你好好好"},
+	}
+
+    // More declarations
+    ...
+
+	// Create a folder for the database file. If a folder by that name exists
+	// already, it will be replaced by an empty folder.
+	testDir := build.TempDir(persistDir, "TestOpenNewDatabase")
+	err := os.MkdirAll(testDir, 0700)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Loop through tests for each testInput.
+	for _, in := range testInputs {
+		dbFilePath := filepath.Join(testDir, in.dbFilename)
+
+		// Create a new database.
+		db, err := OpenDatabase(in.dbMetadata, dbFilePath)
+		if err != nil {
+			t.Fatalf("calling OpenDatabase on a new database failed for input %v; error was %v", in, err)
+		}
+
+		// Close the newly-created, empty database.
+		err = db.Close()
+		if err != nil {
+			t.Fatalf("closing a newly created database failed for input %v; error was %v", in, err)
+		}
+
+        // More tests (check cases where database has already been created)
+        ...
```

Comment:
I suggest using a shorter example of test-driven-development, e.g. siac/parse_test.go. There's a lot of other stuff going on in this test.